firedrake/slate/slac/compiler.py

```diff
@@ -41,18 +45,144 @@
+                    raise NotImplementedError("Subdomains not implemented.")
+
+                args = [c for i in kinfo.coefficient_map
+                        for c in self.coefficient(local_coefficients[i])]
+
+                if kinfo.oriented:
+                    args.insert(0, cell_orientations_sym)
+
+                if kinfo.integral_type in ["interior_facet",
+                                           "exterior_facet",
+                                           "interior_facet_vert",
+                                           "exterior_facet_vert"]:
+                    args.append(ast.FlatBlock("&%s" % it_sym))
+
+                # Assembly calls within the macro kernel
+                tensor = eigen_tensor(exp, self.temps[exp], indices)
+                call = ast.FunCall(kinfo.kernel.name,
+                                   tensor,
+                                   coord_sym,
+                                   *args)
+                assembly_calls[it_type].append(call)
+
+                # Subkernels for local assembly (Eigen templated functions)
+                kast = transformer.visit(kinfo.kernel._ast)
+                templated_subkernels.append(kast)
+                include_dirs.extend(kinfo.kernel._include_dirs)
+                oriented = oriented or kinfo.oriented
+
+        self.assembly_calls = assembly_calls
+        self.templated_subkernels = templated_subkernels
+        self.include_dirs = list(set(include_dirs))
+        self.oriented = oriented
+
+    @property
+    def integral_type(self):
+        """Returns the integral type associated with a Slate kernel."""
+        return "cell"
+
+    @cached_property
+    def needs_cell_facets(self):
+        """Searches for any embedded forms (by inspecting the ContextKernels)
+        which require looping over cell facets. If any are found, this function
+        returns `True` and `False` otherwise.
+        """
+        cell_facet_types = ["interior_facet",
+                            "exterior_facet",
+                            "interior_facet_vert",
+                            "exterior_facet_vert"]
+        return any(cxt_k.original_integral_type in cell_facet_types
+                   for cxt_k in self.context_kernels)
+
+    @cached_property
+    def needs_mesh_layers(self):
+        """Searches for any embedded forms (by inspecting the ContextKernels)
+        which require mesh level information (extrusion measures). If any are
+        found, this function returns `True` and `False` otherwise.
+        """
+        mesh_layer_types = ["interior_facet_horiz_top",
+                            "interior_facet_horiz_bottom",
+                            "exterior_facet_bottom",
+                            "exterior_facet_top"]
+        return any(cxt_k.original_integral_type in mesh_layer_types
+                   for cxt_k in self.context_kernels)
+
+
 def compile_expression(slate_expr, tsfc_parameters=None):
```

Comment:
Even with all the new comments, this function is now very long (almost 350 lines).  It could probably be split into pieces and be easier to follow.

One logical splitting might be to follow the splitting in the generated macro kernel:

1. A pass to declare temporaries
2. A pass to initialise temporaries
3. Generate the linear algebra expression
4. Glue everything together.

WDYT?

Reply:
Thanks for pointing this out. Yeah, I've been wondering if it would be better to split everything up. The outline you just presented is very similar to what I've imagined. Item 3 is already its own method. How about I separate 1 and 2 (possibly combine them together if it can be done cleanly) into their own methods and let `compile_expression` glue everything together?

Reply:
@wence- I went ahead an split `compile_expression` into separate stages. It was primarily code re-factorization. Let me know how you feel about this latest change.